### DISCUSSION TITLE
PyODBC to Oracle 19 Version Connectivity issue on RHEL7 and RHEL8 linux servers in Python virtual environment

### DISCUSSION THREAD

**Discussion by AASupport123** [mkleehammer/pyodbc#1178] [Q&A]
PyODBC to Oracle driver connection is getting failed with the error message 'pyodbc.InterfaceError: ('IM002', '[IM002] [unixODBC][Driver Manager]Data source name not found and no default driver specified (0) (SQLDriverConnect)')' on RHEL7 and RHEL8 linux servers.

We have tried using Oracle 19 version connectivity from python 3.7/3.9 versions using PyODBC on RHEL7 and RHEL8 servers. Please find the below connection code, odbc.ini & tnsnames.ora entry details on server and error message we received when trying the connectivity under virtual environment in python.

Connection Code:
import pyodbc
import os
import pandas as pd
os.environ["ORACLE_HOME"] = "/trvapps/oracle/product/client_64/19c/client"
from ctypes import *
lib1 = cdll.LoadLibrary('/trvapps/oracle/product/client_64/19c/client/lib/libclntsh.so.19.1')
cnxn = pyodbc.connect('DSN=xxx;''uid=xxx;''pwd=xxx;')
cursor = cnxn.cursor()

Error Message:
pyodbc.InterfaceError: ('IM002', '[IM002] [unixODBC][Driver Manager]Data source name not found and no default driver specified (0) (SQLDriverConnect)')

Content in tnsnames.ora file :
EBITS_ora =
  (DESCRIPTION =
    (ADDRESS = (COMMUNITY = TCP)(PROTOCOL = TCP)(Host =xxx)(Port = xxx))
    (CONNECT_DATA =
      (SERVICE_NAME = xxx)
      (SERVER = DEDICATED)
    )
  )

Content in odbc.ini file:                                                     
[EBITS_ora]
Driver=/trvapps/oracle/product/client_64/19c/client/lib/libsqora.so.19.1
DriverUnicodeType=1
Description=Oracle ODBC Database
ServerName=EBITS_ora

**Comment by v-chojas** [comment]
Does `isql -v -k '<your connection string here>'` work?

**Reply by AASupport123** [reply]
Hi v-chojas,

isql -v -k  <connection string> command with oracle connection string is getting failed with same below error on RHEL7 and RHEL8 linux platforms.

isql -v -k 'Driver=oracle;Server=xxx,xxx;Database=EBITS_ora;UID=xxx;PWD=xxx;' --> (Executed command)

Error Message:
[IM004][unixODBC][Driver Manager]Driver's SQLAllocHandle on SQL_HANDLE_HENV failed
[ISQL]ERROR: Could not SQLDriverConnect

**Comment by gordthompson** [comment]
If you want to access an Oracle database from Python you might find it more convenient to use [oracledb](https://pypi.org/project/oracledb/) instead of pyodbc.

**Reply by AASupport123** [reply]
We will try with oracledb and will let you know the test result
